app/blog/tanstack-router-typescript-performance.md

```diff
@@ -0,0 +1,214 @@
+
+<img src="/blog-tracing-faster.png">
+
+The first file to reference a `Link` no longer triggers inference from the whole route tree which increases perceived language service speed significantly
+
+By doing this TypeScript will only infer the types required for a specific route when it is referenced by a `Link`. This may not translate to overall better typescript check time if all routes are used but is significant for the language service
+
+The difference between the two is striking with large route trees with complex inference (400 in this example)
+
+<div style="display: flex">
+  <video src="/language-service-slow.mp4" width="640" height="480" autoplay muted loop></video>
+  <video src="/language-service-fast.mp4" width="640" height="480" autoplay muted loop></video>
+</div>
+
+You maybe thinking that this is cheating, we are indeed making use of a generation step. However this generation step for file-based (and now virtual file based routing) was always there and a necessary step whenever you modify or create a new route. However once the route is created and the route tree is generated, inference remains throughout anything in a route definition. That means if you make any changes to `validateSearch`, `beforeLoad`, `loader` and others the type is inferred and reflected instantly. The DX in this regard has not changed but the performance in the editor can feel awesome for large route trees
+
+## The ground rules
+
+This change has made a lot of the surface of TanStack Router's exports performant for the language service but there are existing areas which will still fallback to forcing inference on the whole route tree. These areas are a kind of loose or non strict modes.
+
+```tsx
+<Link to="." search={{ page: 0 }} />
+<Link to=".." search={{page: 0}} />
+<Link to="/dashboard" search={prev => ({..prev, page: 0 })} />
+```
+
+All three usages of `Link` require inference of the whole route tree and are therefore falling back to a worse editor experience when interacting with them. The first two, TanStack Router does not know what route you want to navigate to and therefore tries its best to guess a very loose type from all routes. The third uses `prev` but TanStack Router does not know what `Route` you are navigating `from` and again tries to guess `prev` with a loose type from all routes. The most performant for your editor is the following:
+
+```tsx
+<Link from="/dashboard" search={{ page: 0 }} />
+<Link from="/dashboard" to=".." search={{page: 0}} />
+<Link from="/users" to="/dashboard" search={prev => ({...prev, page: 0 })} />
+```
+
+TanStack Router can narrow the types to specific routes in these cases. This means that you get better type safety and also better editor performance as your application scales. Therefore we encourage the use of `from` or `to` in these cases. To be clear `from` is only necessary if `prev` is used, otherwise TanStack Router does not infer the whole route tree
```

Comment:
```suggestion
TanStack Router can narrow the types to specific routes in these cases. This means that you get better type safety and better editor performance as your application scales. Therefore, we encourage the use of `from` and/or `to` in these cases. To be clear, in the third example, the usage of `from` is only necessary if the `prev` argument is used, otherwise, TanStack Router does not need to infer the whole route tree.
```